Graphics/PDF/Fonts/Type1.hs

```diff
@@ -52,6 +54,11 @@ getAfmData path = do
     Just r <- parseFont (Right path) 
     return (AFMData r)
 
+parseAfmData :: B.ByteString -> IO AFMData
+parseAfmData bs = do
+    Just r <- parseFont (Left bs)
+    return (AFMData r)
```

Comment:
I don't like these partial functions. I would prefer:

```suggestion
    AFMData <$> parseFont (Right path) 

parseAfmData :: B.ByteString -> IO (Maybe AFMData)
parseAfmData bs = AFMData <$> parseFont (Left bs)
```

Could you update the patch?

Alternatively we could expose `AFMData`'s constructor

Reply:
Thanks for the quick response.

> I don't like these partial functions.

Fully agree here, but I was mirroring getAfmData, which does the same.

```haskell
getAfmData :: FilePath -> IO AFMData 
getAfmData path = do  
    Just r <- parseFont (Right path) 
    return (AFMData r)
```

Looking further into it, it appears even though the types suggest otherwise `parseFont` (in `AfmParser.hs`) calls `error` instead of returning Nothing on failure.

```haskell
parseFont :: Either ByteString String -> IO (Maybe AFMFont)
parseFont (Left bs) = do
    r <- afmParseFromFile afm "<embedded>" bs
    case r of
      Left e -> error (show e)
      Right r' -> return $ Just r'
parseFont (Right path) = do
    bs <- B.readFile path
    r <- afmParseFromFile afm path bs
    case r of
      Left e -> error (show e)
      Right r' -> return $ Just r'
```

I would suggest changing `parseFont` to return the result form `afmParseFromFile` unchanged (possibly exposing `ParseError`) and change both `getAfmData` and `parseAfmData` accordingly:

```suggestion
-- still in AfmParser.hs
parseFont :: Either ByteString String -> IO (Either ParseError AFMFont)
parseFont (Left bs) = afmParseFromFile afm "<embedded>" bs
parseFont (Right path) = do
    bs <- B.readFile path
    afmParseFromFile afm path bs

-- in Type1.hs
import Data.Bifunctor (Bifunctor(second))

getAfmData :: FilePath -> IO (Either ParseError AFMData)
getAfmData = second AFMData . parseFont . Right

parseAfmData :: B.ByteString -> IO (Either ParseError AFMData)
parseAfmData= second AFMData . parseFont . Left
```

> Alternatively we could expose `AFMData`'s constructor

I don't think this would help, since AFMFont isn't exposed anywhere and parseFont actually does some validation.
If it`s only the embedding, then deriving `Lift` would be enough.
However, parsing from `ByteString` can be useful for other purposes as well.

Reply:
> I would suggest changing parseFont to return the result form afmParseFromFile unchanged (possibly exposing ParseError) and change both getAfmData and parseAfmData accordingly:

That would be much better indeed.

Reply:
Looking at it further, `afmParseFromFile` doesn't need to be IO, so `parseAfmData` can be pure as well.